File: resources.rb
require 'json'

class Resource
	@@allres = []
	def initialize (key, type, filename, heavy=false)
		@data = {}
		@data[:key] = key
		@data[:type] = type
		@data[:filename] = filename
		@data[:heavy] = heavy
		@@allres << self
	end
	def getOutput (version)
		resop = {
			:key => @data[:key],
			:type => @data[:type],
			:heavy => @data[:heavy],
		}
		file = File.new(@data[:filename])
		resop[:mtime] = file.mtime.to_i
		if (version < resop[:mtime])
			resop[:content] = file.read
		end
		file.close
		resop
	end
	def self.output(client, params)
		version = params['v']
		if (version.nil? or version.first.nil?)
			version = 0
		else
			version = version.first.to_i
		end
		lite = (params['lite'].first == 'true')
		op = {}
		op[:v] = version
		op[:r] = []
		@@allres.each{ |res|
			resop = res.getOutput(version)
			if (lite and resop[:heavy])
				next
			end
			op[:r] << { :key => resop[:key], :type => resop[:type]}
			if (resop[:content].nil?)
				next
			end
			if resop[:mtime] > op[:v]
				op[:v] = resop[:mtime]
			end
			op[resop[:key]] = resop[:content]
		}
		client.puts("HTTP/1.1 200 OK")
		client.puts
		client.puts(op.to_json)
	end
end

File: resources.js
var fs = require('fs');

var resourcefiles = {
	'_lucos': { filename: __dirname+"/lucos.js", type: 'js'}
};

/**
 * Adds a resource file to the list
 * 
 * @param String key - A unqiue key to identify the file on the client
 * @param String type - tells the client how to parse the file
 * @param String filename - Where in the filesystem the file is stored
 */
exports.add = function (key, type, filename) {
	resourcefiles[key] = {filename: filename, type: type};
}

/**
 * Loads the relevant files
 * 
 * @param ServerResponse res
 * @param String version - The version sent to the client the last time it requested resources
 */
exports.load = function (res, version) {
	var ii;
	var output = {};
	var filenum = 0;
	var totalfiles = 0;
	var vtime = new Date(parseInt(version)*1000);
	var newvtime = vtime;
	var resources = {};
	var isReady = function _isReady() {
		filenum++;
		if (filenum < totalfiles) return;
		output.v = newvtime.getTime() / 1000;
		output.r = resources;
		res.writeHead(200, {'Content-Type': "application/json"});
		res.write(JSON.stringify(output));
		res.end();
	}
	for (ii in resourcefiles) {
		(function _addFileToOutput(key, filename, type) {
			resources[key] = type;
			fs.stat(filename, function _gotstat(err, stats) {
				if (err) {
					isReady();
					return;
				}
				var mtime = new Date(stats.mtime);
				if (mtime > vtime) {
					if (mtime > newvtime) newvtime = mtime;
					fs.readFile(filename, function _gotfile(err, data) {
						if (err) {
							isReady();
							return;
						}
						output[key] = data.toString();
						isReady();
					});
				} else {
					isReady();
				}
			});
		})(ii, resourcefiles[ii].filename, resourcefiles[ii].type);
		totalfiles++;
	}
}
